### ISSUE TITLE
BertForTokenClassification support

### ISSUE THREAD

**Issue by jmriebold** [aws/aws-neuron-sdk#304]
Hey, I'm interested in converting a Transformers [BertForTokenClassification](https://huggingface.co/transformers/model_doc/bert.html#bertfortokenclassification) model I use to Neuron so that I can deploy to inf1 instances, and although I'm able to compile a Neuron model using a modified version of this [tutorial](https://github.com/aws/aws-neuron-sdk/blob/master/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.ipynb), the model's predictions are completely different from the PyTorch model. I noticed during compilation that the model's embedding layer is not supported. Looks like support for embedding layers was [removed](https://github.com/aws/aws-neuron-sdk/blob/master/release-notes/neuron-cc-ops/neuron-cc-ops-pytorch.rst#pytorch-neuron-release-1010010) some time ago. Is this model type not supported?

Happy to attach an example if this is not expected behavior.

**Comment by micwade-aws** [comment]
Hi @jmriebold - 

The embedding layer will be run on CPU rather than Inferentia since it is not yet supported by the SDK. This only means the embedding layer will not be accelerated in any way.

Could you attach your example code so we can try to reproduce the behavior you are seeing?

**Comment by aws-joshim** [comment]
@jmriebold checking if we can further assist with any questions on the BertForTokenClassification https://huggingface.co/transformers/model_doc/bert.html#bertfortokenclassification model on Neuron?

**Comment by jmriebold** [comment]
Hey @aws-joshim, apologies for the delay. I'm gonna try and put together some example code showing the issue today.

**Comment by jmriebold** [comment]
Ok, so first, @micwade-aws, thanks for the clarification on unsupported layers. Second, I spent a couple hours working on this today putting together an example script that I thought was roughly equivalent to what we were doing, but it turns out I can't replicate the issue with a pre-trained model.

We're using a BertForTokenClassification model based on bert-base-cased fine-tuned with our own data. The labels appear to be ordered differently from the standard set used with dbmdz/bert-large-cased-finetuned-conll03-english (which is what I used for my testing today), but other than that I'm not sure what's different (unfortunately I wasn't the one who trained the model).

To give you a little more detail on what I'm seeing, I came up with some example text which uses all of the different entity types the model knows about (since the resulting Neuron model seemed to be affected by both the length of the example and the types of entities in it):

> Morningstar, Inc. (NASDAQ: MORN), a leading provider of independent investment research, today announced that it has entered into a definitive agreement to acquire PitchBook Data, Inc. PitchBook, founded in 2007 in Seattle, WA, delivers data, research, and technology covering  the breadth of the private capital markets, including venture capital, private equity, and mergers and acquisitions (M&A). PitchBook will maintain its brand and identity and will continue to be led by founder and chief executive officer John Gabbert. Observers noted that the Securities Exchange Act of 1934 is very important.

Then ran predictions with the existing Torch model:
```
[CLS] O
Morning B-Organization
##star I-Organization
, I-Organization
Inc I-Organization
. I-Organization
( O
NAS B-Organization
##DA I-Organization
##Q I-Organization
: O
M B-Organization
##OR B-Organization
##N I-Organization
) O
, O
a O
leading O
provider O
of O
independent O
investment O
research O
, O
today O
announced O
that O
it O
has O
entered O
into O
a O
definitive O
agreement O
to O
acquire O
Pit B-Organization
##ch I-Organization
##B I-Organization
##ook I-Organization
Data I-Organization
, O
[SEP] O
[CLS] O
Inc B-Organization
. I-Organization
Pit B-Organization
##ch I-Organization
##B I-Organization
##ook I-Organization
, O
founded O
in O
2007 O
in O
Seattle B-Location
, O
WA B-Location
, O
delivers O
data O
, O
research O
, O
and O
technology O
covering O
the O
bread O
##th O
of O
the O
private O
capital O
markets O
, O
including O
venture O
capital O
, O
private O
equity O
, O
and O
merger O
##s O
and O
acquisitions O
( O
M O
& O
A O
) O
. O
[SEP] O
[CLS] O
Pit B-Organization
##ch I-Organization
##B I-Organization
##ook I-Organization
will O
maintain O
its O
brand O
and O
identity O
and O
will O
continue O
to O
be O
led O
by O
founder O
and O
chief O
executive O
officer O
John B-Person
G I-Person
##ab I-Person
##bert I-Person
. O
[SEP] O
[CLS] O
Observer O
##s O
noted O
that O
the O
Securities B-Miscellaneous
Exchange I-Miscellaneous
Act I-Miscellaneous
of I-Miscellaneous
1934 I-Miscellaneous
is O
very O
important O
. O
[SEP] O
```

then ran the conversion script using the tutorial code (only modifications were some custom text normalization and sentence breaking, along with the use of our own model), and reran predictions:

```
[CLS] O
Morning B-Organization
##star I-Organization
, I-Organization
Inc I-Organization
. I-Organization
( O
NAS B-Organization
##DA I-Organization
##Q I-Organization
: O
M B-Organization
##OR B-Organization
##N I-Organization
) O
, O
a O
leading O
provider O
of O
independent O
investment O
research O
, O
today O
announced O
that O
it O
has O
entered O
into O
a O
definitive O
agreement O
to O
acquire O
Pit B-Organization
##ch I-Organization
##B I-Organization
##ook I-Organization
Data I-Organization
, I-Organization
[SEP] I-Organization
[CLS] O
Inc B-Organization
. I-Organization
Pit I-Organization
##ch I-Organization
##B I-Organization
##ook O
, B-Organization
founded I-Organization
in I-Organization
2007 O
in B-Organization
Seattle B-Organization
, I-Organization
WA O
, O
delivers O
data O
, O
research O
, O
and O
technology O
covering O
the O
bread O
##th O
of O
the O
private O
capital O
markets O
, O
including O
venture O
capital O
, B-Organization
private I-Organization
equity I-Organization
, I-Organization
and I-Organization
merger I-Organization
##s I-Organization
and I-Organization
acquisitions B-Organization
( I-Organization
M I-Organization
& I-Organization
A O
) O
. O
[SEP] O
[CLS] O
Pit B-Organization
##ch I-Organization
##B I-Organization
##ook I-Organization
will I-Organization
maintain O
its B-Organization
brand I-Organization
and I-Organization
identity O
and B-Organization
will B-Organization
continue I-Organization
to O
be O
led O
by O
founder O
and O
chief O
executive O
officer O
John O
G O
##ab O
##bert O
. O
[SEP] O
[CLS] O
Observer B-Organization
##s I-Organization
noted I-Organization
that I-Organization
the I-Organization
Securities O
Exchange B-Organization
Act I-Organization
of I-Organization
1934 O
is B-Organization
very B-Organization
important I-Organization
. O
[SEP] O
```

As you can see, the predictions are completely different, but I'm a little out of my depth here and am unsure of how to proceed in terms of troubleshooting. For reference, here's our model.json:
```
{
  "_name_or_path": "bert-base-cased",
  "architectures": [
    "BertForTokenClassification"
  ],
  "attention_probs_dropout_prob": 0.1,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "O",
    "1": "B-Organization",
    "2": "I-Organization",
    "3": "B-Location",
    "4": "I-Location",
    "5": "B-Person",
    "6": "I-Person",
    "7": "B-Miscellaneous",
    "8": "I-Miscellaneous"
  },
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "label2id": {
    "B-Location": 3,
    "B-Miscellaneous": 7,
    "B-Organization": 1,
    "B-Person": 5,
    "I-Location": 4,
    "I-Miscellaneous": 8,
    "I-Organization": 2,
    "I-Person": 6,
    "O": 0
  },
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "torch_dtype": "float32",
  "transformers_version": "4.9.0",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 28996
}
```

**Comment by nonatofabio** [comment]
Hi @jmriebold,
Thanks for sharing the results. Regarding the prediction comparison, running test over a complete dataset will give you better metrics for accuracy than comparing a single inference prediction. Single predictions of a compiled model can differ from a non-compiled model, due to the auto-casting feature of Neuron.  However, the accuracy over a test dataset should be minimally impacted. Evaluating the accuracy impact over a full NER test dataset, i.e conll03, is our recommended next step in terms of troubleshooting.

**Comment by jmriebold** [comment]
Thanks @nonatofabio, I'll give that a try next week.

**Comment by jmriebold** [comment]
Hi @nonatofabio, I wasn't able to find any existing scripts for running Neuron models against test sets that wouldn't need significant modifications, but I did run 20 more examples from our test set and CONLL2003 through the Neuron model, and nearly all of the predictions were wrong. Similar to the above examples, all words were marked as either `Organization` or `O`.

This suggests to me that either something has gone very wrong in the model conversion process, or there's something custom being done in our NER code that isn't getting carried over to the Neuron model. I've taken a look at the training code, but it looks like a very standard usage of the Transformers library so far as I can tell, nothing that isn't already in the config.json I posted above.

**Comment by aws-zejdaj** [comment]
@jmriebold I suggest to pursue the following debug paths:
1. Please try to use Neuron compile option --fp32-cast=matmult  (If not used in your compile scripts already )
2. Confirm that you get good classification accuracy when you compile your new script on cpu
3. Share your exact script/model - here on github or reach to us directly at aws-neuron-support@amazon.com

**Comment by jmriebold** [comment]
Thanks a lot @aws-zejdaj, I'll give that a try!

**Comment by jmriebold** [comment]
Just gave this a try, results look identical to the models I trained without the arg.

The command being executed is this, in case you're curious: `/home/ubuntu/anaconda3/envs/aws_neuron_pytorch_p36/bin/neuron-cc compile /tmp/tmpizaaxvgx/graph_def.pb --framework TENSORFLOW --pipeline compile SaveTemps --output /tmp/tmpizaaxvgx/graph_def.neff --io-config {"inputs": {"0:0": [[1, 128, 768], "float32"], "1:0": [[1, 1, 1, 128], "float32"]}, "outputs": ["add_132:0"]} --fp32-cast=matmult --verbose 35'`

**Comment by jeffhataws** [comment]
Thanks for the valuable input. We are taking a look.

**Comment by mrnikwaws** [comment]
Hi @jmriebold,

We were able to reproduce the problem you are seeing with BertForTokenClassification and have identified the fix in Neuron, it will be available in a future release

**Comment by jmriebold** [comment]
@mrnikwaws that's great news! Really appreciate you taking the time to investigate.

**Comment by aws-diamant** [comment]
Hi @jmriebold,

Just a quick update that this issue is expected to be solved by this [roadmap item](https://github.com/aws/aws-neuron-sdk/issues/333), which will allow developers to have better control over accuracy vs speed tradeoff.
This is already implemented, and will be available in an upcoming release.

**Comment by jmriebold** [comment]
Awesome, I'll make sure to follow that issue. Thanks @aws-diamant, looking forward to trying it out when it's ready!

**Comment by awsrjh** [comment]
closing based on https://github.com/aws/aws-neuron-sdk/issues/333 being available now

Please see: 

https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/appnotes/perf/mixed-precision.html#compiler-casting-options
